### DISCUSSION TITLE
Saving mesh to EEPROM question

### DISCUSSION THREAD

**Discussion by adam3654** [Jyers/Marlin#1048] [Q&A] [answered]
One of the little mods I made in my own copy of the firmware is to automatically save the newly created mesh to EEPROM.

When I run a new mesh I walk away and let it do it's thing, then would always forget to walk back down to the basement and click 'confirm' to save the mesh. 

If one is running a new mesh it is because they want to change the existing mesh, and if it comes out wonky, they will be running another new one anyways. 

I have always wondered if there was ever a reason NOT to save it.

**Comment by smooreace** [comment, marked as answer]
I personally would think that the natural progression for the process would
be..

Start the mesh sample (should include bed and nozzle heating)
Display mesh results when finished with a Resample, Save, and Cancel button
at the end.

**Comment by Jyers** [comment]
Honestly, the most likely reason is time. Some people use the mesh to level the bed, or have to run it multiple times to get it accurate. With a 15x15 bed mesh, it can take quite a few seconds to save the mesh to eeprom. Some would just prefer to skip that, check the mesh first, and then save it if it's good.

It is also important if you are using multiple meshes, that way if at the end you're not quite sure you have the right mesh slot set, you can choose to not save and check it first. Otherwise you could easily accidentally overwrite the wrong mesh.

**Reply by smooreace** [reply]
The biggest thing that bugs me is having to wait for a 10x10 BLT "scan" to complete (IF it actually completes) only to have conflicting final results and have to jump through menus just to view it. My current example.. graphically I see variations between  - .01 and -.44. Thats like... beyond not a thing. It should be taking the average of the min and max samples and calculating all points as a + or - from that average. One cant have an all negative sample... it defies logic!

**Reply by Jyers** [reply]
Although that may seem illogical, it makes sense in the scope of how the printer is operating. All offsets are based off of what the printer knows to be zero. And all it knows of zero is the point in the middle where it probes to home, therefore the mesh points are going to be based on their offset from that point. In your case it sounds like the middle of your bed is higher than the rest of it, causing all the points to be negative offsets.

**Reply by Jyers** [reply]
I do like the idea of viewing the mesh in the save popup though, I'll think over that one.

**Reply by adam3654** [reply]
@smooreace I am probing with the BL Touch, so I am only heating the bed (enabled PREHEAT_BEFORE_PROBING) then I added code to set the Bed temp back to 0 after).  Also I was just checking the marlin g-code reference and noticed this:

`P1 will suspend mesh generation if the controller button is held down. Note that you may need to press and hold the switch for several seconds if moves are underway.`

@Jyers  does this carry over to the new UI?

When I am creating a new mesh it is because I have reason to believe the bed "shape" has changed (i.e. new glass bed, moved printer around to work on it, etc.), therefore the old mesh is no longer useful, and not worth keeping. I think auto saving the mesh works better for me in my use case, I only utilize slot 0, and I would rather limit my walks back and forth to the basement. ;)

The mesh view screen would be sweet, but could the individual grid point data be returned to the display point by point? G29 source could be overhauled probably, but wouldn't that go against the rules (leaving the Marlin stuff alone)?

**Reply by smooreace** [reply]
I suppose I can see that. It's just been a frustrating battle for a week to hone in on as close to a level starting point as possible. And part of that is all the menu hoping that's required.

**Reply by Jyers** [reply]
> The mesh view screen would be sweet, but could the individual grid point data be returned to the display point by point?

Have you never used the mesh viewer in the leveling menu? The code is already there to get and plot out the mesh points, the new thing would be putting it in a popup.

> @Jyers does this carry over to the new UI?

No that is a MarlinUI mechanic and is not present in the code currently.

**Reply by adam3654** [reply]
I guess I was thinking more in real-time as the G29 P1 is reading the points, it could be populating the viewer as the points are measured, while displaying the min/max/avg values, and maybe for a little extra razzle-dazzle toss a cancel button at the bottom.

**Reply by Jyers** [reply]
Ohhh... That would be cool if it's possible!

**Reply by adam3654** [reply]
I see if EMERGENCY_PARSER is enabled, we might be able to pump out some info a la:

```
/**
 * Emergency Command Parser
 *
 * Add a low-level parser to intercept certain commands as they
 * enter the serial receive buffer, so they cannot be blocked.
 * Currently handles M108, M112, M410, M876
 * NOTE: Not yet implemented for all platforms.
 */
//#define EMERGENCY_PARSER

/**
 * Realtime Reporting (requires EMERGENCY_PARSER)
 *
 * - Report position and state of the machine (like Grbl).
 * - Auto-report position during long moves.
 * - Useful for CNC/LASER.
 *
 * Adds support for commands:
 *  S000 : Report State and Position while moving.
 *  P000 : Instant Pause / Hold while moving.
 *  R000 : Resume from Pause / Hold.
 *
 * - During Hold all Emergency Parser commands are available, as usual.
 * - Enable NANODLP_Z_SYNC and NANODLP_ALL_AXIS for move command end-state reports.
 */
//#define REALTIME_REPORTING_COMMANDS
#if ENABLED(REALTIME_REPORTING_COMMANDS)
  //#define FULL_REPORT_TO_HOST_FEATURE   // Auto-report the machine status like Grbl CNC
#endif
```

I am at work, so I can't deep dive in the code to find the destination of the `FULL_REPORT_TO_HOST_FEATURE` but it would be interesting if that could be utilized without modifying the g-code source file.

**Reply by Jyers** [reply]
Oh I'm fairly sure it's possible without need to modify the main marlin source files. Leveling procedures should save values to the z values matrix even during probing, I mean where else would they put them? So ideally it would be easy enough to just grab those as it goes and display them appropriately. 

The emergency parser is really for communications with a host device, not onboard devices like the LCD

**Reply by Jyers** [reply]
I'm also at work, so perhaps I'll look into it a bit when I get home to see if they do get saved to the matrix like I'd expect
